Add auto-margin scaling in draw_networkx_nodes function (fix for issue 3443)

EricPostMaster opened the pull request:
<!--
Please run black to format your code.
See https://networkx.org/documentation/latest/developer/contribute.html for details.
-->

Pull request addressing [issue 3443](https://github.com/networkx/networkx/issues/3443)

- [x] Run Black
- [x] Passed tests

dschult commented on networkx/drawing/nx_pylab.py:
```diff
@@ -470,6 +476,40 @@ def draw_networkx_nodes(
+        try:  # sets x and y margins separately if node_size is iterable
+            node_coords_sizes = list(zip(pos.items(), node_size))
+
+            x_min = sorted(node_coords_sizes, key=lambda a: a[0][1][0], reverse=False)[
+                0
+            ]
+            x_max = sorted(node_coords_sizes, key=lambda a: a[0][1][0], reverse=True)[0]
```
```suggestion
            sortx = sorted(node_coords_sizes, key=lambda a: a[0][1][0])
            x_min = sortx[0]
            x_max = sortx[-1]
```
Saves sorting twice if you use first and last result of sort.
It might be better to use `min` and `max` than sort -- not sure.
The same suggestion applies to `y_min` and `y_max`.

dschult commented on networkx/drawing/nx_pylab.py:
```diff
@@ -470,6 +476,40 @@ def draw_networkx_nodes(
+            x_max = sorted(node_coords_sizes, key=lambda a: a[0][1][0], reverse=True)[0]
+            x_node_size = max([x_min, x_max], key=itemgetter(1))[1]
+
+            if x_node_size >= 500:
```
Can you add comments explaining why the constants here are 100, 72, 50 and maybe 500 as well. If arbitrary or based on what seems to work well, then say that too.

dschult reviewed (commented):
Thanks for this!  
I think the next thing we need to do if find the corner cases that this doesn't work well for.

rossbar reviewed (commented):
Just my two cents, but the custom scaling seems overly complicated. How would others feel about an alternative of providing a `margins` kwarg instead that passes the value directly through to `ax.margins`. On one hand, such argument passthrough is inelegant but is an established pattern in the way `nx_pylab` currently works, and introduces far less untested code than the autoscaling does. The main advantage of adding this kwarg would be a) we could set a different default value than matplotlib's default and b) it raises the visibility of this option to the users by having an explicitly documented parameter.

The code is very nice (thanks for writing it out), I just prefer a simpler solution if possible!

dschult commented:
After some thought I agree with @Rossbar. 
Let's keep it simple by adding a new optional argument `margins` that (if specified) gets passed to `ax.margins`.
:)

EricPostMaster commented:
That works. @dschult, do you want me to go ahead and make that update on this pull request?

dschult commented:
Yes... that would be helpful. Thanks!

EricPostMaster commented:
@dschult I removed the autoscaling code and added the optional `margins` argument. It's not as fun as the autoscaling, but it's functional and much simpler, so the Zen of Python would approve :)

EricPostMaster commented:
I don't understand what is causing these test errors. Everything worked just fine when I pushed before, and now the code is even simpler but returning errors. @dschult or @rossbar, can you help me understand what's happening here so I can figure out how to fix it?

dschult commented on networkx/drawing/nx_pylab.py:
```diff
@@ -470,6 +474,8 @@ def draw_networkx_nodes(
         labelleft=False,
     )
 
+    ax.margins(margins[0], margins[1])
+
```
```suggestion
    if margins is not None:
        ax.margins(margins[0], margins[1])

```
I think this is the cause of (at least) most errors.
We have to treat the default case explicitly.

dschult commented on networkx/drawing/nx_pylab.py:
```diff
@@ -402,6 +403,9 @@ def draw_networkx_nodes(
     label : [None | string]
         Label for legend
 
+    margins : [None | list] (default=None)
+        Horizontal and vertical plot margins. List should contain exactly two elements.
```
```suggestion
    margins : Sequence of 2 numbers or None (default=None)
        The sequence contains horizontal and vertical axis margins. Adjust to avoid image being clipped.
```

dschult reviewed (commented):
I think the errors are only for the default case. 
We'll see if others still appear.

rossbar commented:
> the easiest way would just be to get rid of the unpacking and switch to a pure pass-through, i.e. ax.margins(margins).

Hmm, looking more closely at my own suggestion, this might not be enough since MPL implements this parsing in the signature, whereas we only have a single kwarg to work with. So to support both scalar and tuple inputs we may actually have to add a bit of parsing logic of our own (darn), e.g.

```python
ax.margins(*margins) if isinstance(Iterable, margins) else ax.margins(margins)
```

or something similar (a try/except, etc.).

EricPostMaster commented:
@rossbar - Thanks, I added that to the function, with a slight adjustment switching `margins` and `Iterable`:
```python
ax.margins(*margins) if isinstance(margins, Iterable) else ax.margins(margins)
```

rossbar commented on networkx/drawing/nx_pylab.py:
```diff
@@ -402,6 +403,9 @@ def draw_networkx_nodes(
     label : [None | string]
         Label for legend
 
+    margins : Sequence of 2 numbers or None (default=None)
+        The sequence contains horizontal and vertical axis margins. Adjust to avoid image being clipped.
```
The description here should be modified to reflect the changes below, maybe something like:

```suggestion
    margins : float or 2-tuple, optional
        Sets the padding for axis autoscaling. This can be helpful in cases
        where nodes are clipped near the edges of an image - increasing the
        margin prevents clipping. Values should be in the range ``[0, 1]``.
        see :meth:`matplotlib.axes.Axes.margins` for details. The default
        is `None`, which uses the Matplotlib default.
```

EricPostMaster replied:
Made a small edit to the docstring you suggested to shorten it a bit. 👍

rossbar reviewed (commented):
Generally LGTM, we just have to make sure the docstring accurately describes what the inputs can be, i.e. either float or 2-tuple of floats (or None, which is usually denoted by `optional`).

Another thing is whether we want to make the `margins` kwarg work with `nx.draw` and `nx.draw_networkx`. I think this can be done by adding `"margins"` to the `valid_node_kwds` list: 

https://github.com/networkx/networkx/blob/01d9cfe028e82ce7f2f7fd378ffe6e43ba2fddbb/networkx/drawing/nx_pylab.py#L266-L279

It may be more involved though so if you don't want to worry about it here that's fine! I'm +1 for merge w/ a docstring update.

rossbar reviewed (approved):
LGTM, thanks @EricPostMaster!

dschult commented on networkx/drawing/nx_pylab.py:
```diff
@@ -470,6 +477,9 @@ def draw_networkx_nodes(
     )
 
+    if margins is not None:
+        ax.margins(*margins) if isinstance(margins, Iterable) else ax.margins(margins)
```
I'm not sure if standards have changed, but I think python zen says we use the in-line `___ if ___ else ___` when the result produces an object (that is, when the in-line feature is what is desired).  For flow control we should use usual `if` structure...  I don't feel strongly, but what do you think?

```suggestion
        if isinstance(margins, Iterable):
            ax.margins(*margins)
        else:
            ax.margins(margins)
```

rossbar replied:
I wasn't aware of that - sounds like a good suggestion to me :+1:

EricPostMaster replied:
That's an interesting point that I was not aware of either. Sounds good!

dschult reviewed (approved):
Thanks for this!

rossbar commented:
Ah - sphinx 4.0 came out today :). This is almost certainly a result of that - there's nothing in this PR that would cause this particular error, which I've never seen before either.

dschult commented:
Is this another case where we should be testing against prerelease code? Or is that even possible with sphinx_build?
:{

rossbar commented:
>  If it is OK with you @rossbar maybe we should merge despite a failing test.

:+1: definitely!

rossbar commented:
> Is this another case where we should be testing against prerelease code? Or is that even possible with sphinx_build?

It's possible, but I think projects will need to decide which "core" dependencies to test against their dev branches, otherwise things could get very noisy - I'm not sure what the best-practice would be here :)

Another thing this reminds me is that we should probably pin the upper bound on the major version number of every one of our dependencies. The only reason I've been hesitant to do that is because I don't want to fall into a situation where we accidentally don't upgrade to a new major version of a dependency because we've forgotten about a particular pin. There are tools like dependabot to help with that specifically, but I think that one in particular is a bit noisy. Again - I'm not sure what the best-practice is!

rossbar commented:
> Ah - sphinx 4.0 came out today :). This is almost certainly a result of that

I spoke too soon - we are properly pinned to Sphinx 3.5.4 so I was hasty in attributing this to Sphinx 4. Either way - still +1 for merging as I'm confident the circleCI error is not related to this PR.
